Challenge Stablecoins

escottalexander opened the pull request:
This PR will add the stablecoin challenge to the se-2-challenges repo.

Please check it over and let me know if there are any ways we can improve the learning experience.

Here is the completed contract: https://github.com/escottalexander/stablecoin-challenge/blob/main/packages/hardhat/contracts/MyUSDEngine.sol

The main hero image in the readme will not work until the branch is merged.

Thank you.

escottalexander commented:
@rin-st Should I go ahead and update this to work with the latest version of create-eth? The only issue I know of currently is that the tailwind styles will not be applied. I would need to look into how to make it work and I suppose at some point we will need to do that with all the challenges. Any thoughts on this?

rin-st commented:
Hey, thank you for a awesome challenge! Got some a-ha moments of how the system works. Great job!

---

Some comments:

- Change user row color in tables, theme, font and create-eth version (0.2.4 for now) like in this PR https://github.com/scaffold-eth/se-2-challenges/pull/297/files

- I think `yarn deploy —reset` could be mentioned before this line too:
`🧪 Go test the minting functionality on the front end. After depositing collateral, hover the mint icon and input the amount of MyUSD you would like to mint.`

- In the `withdrawCollateral(uint256 amount)` solution Event `CollateralWithdrawn` has 4 args instead of 3

- Price graph doesnt change when clicking  on show/hide rates button (just width changes)

And some comments using example. 

Get eth from faucet 1000000 (1M) Eth
Collateral 500000 (500K) Eth
Mint 500M MyUsd. Circulating and Staked on the chart became 500M. But I didn’t stake anything, so staking should be 0?
Sell (Swap) 10M MyUsd to Eth. Price graph starts from this point. But maybe it should start from 1$ on the previous step (after mint)?
Stake 400M MyUsd. Chart  shows Staking at 490M (500M - 10M), but it should be 400M?
Sell (Swap) 40M MyUsd for Eth
Buy (Swap) 50M MyUsd. Last two steps changes price and also both staking and circulating. But again looks like staking should remain the same

Left arrow: price start, right arrow: staking value
![image](https://github.com/user-attachments/assets/ddb52d0f-525e-4a4f-b1fc-d6d07ebff030)

Thank you!

AzimovS commented:
Hi @Pabl0cks ! Thank you very much for testing and feedback! What do you think—where can we display the deposited ETH as collateral? We decided to remove it because there wasn’t enough space in the table.

Pabl0cks commented:
> Maybe we could abbreviate it and show actual numbers when below 9.99 and then show only 10, and when we enter the thousands just show 10k, 1.2m etc? 

This sounds good!

We can show the coin name (ETH / MyUSD) in the table header titles

escottalexander commented:
@rin-st @Pabl0cks - Good catches and good suggestions all around. Thanks so much for your reviews. @AzimovS and I have addressed all of your feedback. Let us know if you run into any other issues.

escottalexander commented:
> We can show the coin name (ETH / MyUSD) in the table header titles
Azimov is displaying the asset type on hover since it takes too much space to show it in the header. Also, the student should be able to realize that the Collateral is always ETH and the debt is always MyUSD.

rin-st reviewed (commented):
Great job guys! Sorry for slow response! For some reason I thought all my comments are resolved

> Sell (Swap) 10M MyUsd to Eth. Price graph starts from this point. But maybe it should start from 1$ on the previous step (after mint)?

This still works same way as before, but it's mostly nitpick so could be ignored

> Price graph doesnt change when clicking on show/hide rates button (just width changes)

This is fixed, thanks!

> What about 3 values: total, staked, circulating? And value on hover? 

It's much better now! 
<img width="484" alt="Screenshot 2025-07-01 at 14 54 32" src="https://github.com/user-attachments/assets/451aa42d-a8b1-47fd-93ee-d35055660b25" />

but it doesnt show Circulating chart for me, only Total and Staked

escottalexander commented on README.md:
```diff
@@ -1,6 +1,1118 @@
+
+🛠️ Before you begin, you need to install the following tools:
+
+- [Node (v22 LTS)](https://nodejs.org/en/download/)
```
I will change this. Thank you.

escottalexander commented on README.md:
````diff
@@ -1,6 +1,1118 @@
+📥 Then download the challenge to your computer and install dependencies by running:
+
+```sh
+npx create-eth@0.1.8 -e scaffold-eth/se-2-challenges:challenge-stablecoins challenge-stablecoins 
````
Confirmed that 0.2.4 works properly with styles populated. Latest commit updates the readme.

escottalexander commented:
@rin-st The fix for the thing you mentioned has been merged into this branch. Thanks for pointing it out, Rinat!

Pabl0cks reviewed (commented):
Good job with the tweaks 🙌 I think the UI is working nicely now.

I’ll review the README more carefully tomorrow morning 🙏

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1118 @@
+![readme-stablecoin](https://raw.githubusercontent.com/scaffold-eth/se-2-challenges/challenge-stablecoins/extension/packages/nextjs/public/hero.png)
 
-There are `// CHALLENGE-TODO:` comments placed throughout that provide instructions on key changes to adapt this template for use with a new challenge.
+> 🪙 Build your own decentralized stablecoin! In this challenge, you'll build the core engine for **MyUSD**, a crypto-backed stablecoin designed to maintain a peg to $1 USD. You'll get to wear the hat of a DeFi protocol that wants to maintain price stability while also increasing adoption of your stablecoin product, diving deep into concepts like collateralization, minting, burning, interest rates ...
```
Nit: Maybe I'd remove the citation to match the other introduction paragraphs from other challenges, and split in two paragraphs.

```suggestion
🪙 Build your own decentralized stablecoin! In this challenge, you'll build the core engine for **MyUSD**, a crypto-backed stablecoin designed to maintain a peg to $1 USD.

You'll get to wear the hat of a DeFi protocol that wants to maintain price stability while also increasing adoption of your stablecoin product, diving deep into concepts like **collateralization, minting, burning, interest rates,** and **liquidations** – all crucial components of a robust stablecoin system.
```

Pabl0cks commented on README.md:
````diff
@@ -1,6 +1,1118 @@
+📥 Then download the challenge to your computer and install dependencies by running:
+
+```sh
+npx create-eth@0.2.4 -e scaffold-eth/se-2-challenges:challenge-stablecoins challenge-stablecoins 
````
Nit: remove last space?

```suggestion
npx create-eth@0.2.4 -e scaffold-eth/se-2-challenges:challenge-stablecoins challenge-stablecoins
```

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1118 @@
+
+5. 📊 **Oracle (`Oracle.sol`)**
+   - Provides ETH/MyUSD and ETH/USD price feeds
+   - ETH/USD price is **fixed** at time you deploy the contracts
```
nit: at the time

```suggestion
   - ETH/USD price is **fixed** at the time you deploy the contracts
```

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1118 @@
+
+---
+
+🚀 Go ahead and re-deploy your contracts with `yarn deploy --reset` and test your front-end to see if you can add collateral. 
```
```suggestion
🚀 Go ahead and re-deploy your contracts with `yarn deploy --reset` and test your front-end to see if you can add collateral.
```

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1118 @@
+
+🛠️ Now let's set up the ability for the rate controller to change the borrow rate.
+
+Whenever the rate is changed we need to "lock-in" all the interest accrued since the last rate change using the `_accrueInterest` method we created in checkpoint 2.
```
Should it be "Checkpoint 3"?

escottalexander commented on README.md:
```diff
@@ -1,6 +1,1118 @@
+
+---
+
+Keep in mind, in the absence of decimals we will assume that a borrow rate of 125 is equivalent to a 1.25% annual rate. This will mean we need to multiply by 10000
```
Nice catch. I think I forgot to finish that sentence. Here is what I am changing it to:
"This will mean we need to divide by 10000 (i.e. 100.00%) any time we have multiplied by the borrow rate."

escottalexander commented:
@rin-st @Pabl0cks We (@AzimovS and I) have addressed all your feedback up to this point. Thanks so much for the detailed examination you both put into this challenge. Let us know if you notice anything else.

rin-st reviewed (approved):
> We opted to just show the two values as you suggested, Rinat:

Thank you! Lgtm! Great job!
Please wait for Pablo's approve and it will be ready to merge

Pabl0cks reviewed (approved):
LGTM!! Good job everyone! 🔥

escottalexander commented:
When this is merged we can update the autograder: https://github.com/austintgriffith/speedrun-grader/pull/17

Just don't want to forget. :)

rin-st commented:
Rerun the tests with a contract from the initial message, all tests passed

Thank you, merging this!

cc @carletex

carletex commented:
sorry, busy day!

I'll update the autograder tomorrow.

Thanks all!
